rd-kt/rd-core/src/main/kotlin/com/jetbrains/rd/util/collections/SynchronizedSet.kt

```diff
@@ -0,0 +1,76 @@
+package com.jetbrains.rd.util.collections
+
+import java.util.function.Predicate
+
+class SynchronizedSet<T> : MutableSet<T> {
+    companion object {
+        private val PRESENT = Any()
+    }
+
+    private val map = SynchronizedMap<T, Any>()
+
+    override fun add(element: T): Boolean {
+        return map.put(element, PRESENT) == null
+    }
+
+    override fun addAll(elements: Collection<T>): Boolean {
+        var added = false
+        elements.forEach {
+            if (add(it))
+                added = true
+        }
+
+        return added
+    }
+
+    override val size: Int
+        get() = map.size
+
+    override fun clear() {
+        map.clear()
+    }
+
+    override fun isEmpty(): Boolean {
+        return map.isEmpty()
+    }
+
+    override fun containsAll(elements: Collection<T>): Boolean {
+        return map.keys.containsAll(elements)
+    }
+
+    override fun contains(element: T): Boolean {
+        return map.containsKey(element)
+    }
+
+    override fun iterator(): MutableIterator<T> {
+        val iterator = map.iterator()
+
+        return object : MutableIterator<T> {
+            override fun hasNext(): Boolean = iterator.hasNext()
+            override fun next(): T = iterator.next().key
+            override fun remove() = iterator.remove()
```

Comment:
I think we should support as many methods as possible, if it's not too complicated. Because we plan to use these collections as backing implementations for protocol entities, and you'll never know which methods were used in the production code)